hy/compiler.py

```diff
@@ -1709,10 +1722,22 @@ def hy_compile(tree, module_name, root=ast.Module, get_expr=False):
 
     tree = wrap_value(tree)
     if not isinstance(tree, HyObject):
-        raise HyCompileError("`tree` must be a HyObject or capable of "
-                             "being promoted to one")
-
-    compiler = HyASTCompiler(module_name)
+        raise TypeError("`tree` must be a HyObject or capable of "
+                        "being promoted to one")
+
+    try:
+        mod_loader = pkgutil.get_loader(module_name)
+        if filename is None:
+            filename = mod_loader.get_filename()
+        # TODO: Lazy load this.
+        if source is None:
+            source = mod_loader.get_source()
+    except Exception:
```

Comment:
Again, `except Exception` might be too broad.

Reply:
What I'm afraid of is silencing an error that would alert us of a bug. That seems worse to me than weird crashes from unanticipated but meaningless errors we should've ignored. I'd say, just list the exception types you've been able to provoke so far. We'll add more later if we need them. Does that sound reasonable?

Reply:
Any relevant errors showing up here would necessarily appear in more appropriate places soon after (i.e. when one of the modules is actually imported), or&mdash;best case scenario&mdash;we would completely avoidable irrelevant errors in the core functionality that follows.

I don't mind being more precise, though, so I'll make that change.

Reply:
Great example of my point: the current `pytest` failures are due to an exception in this exact place (and the exact circumstances I described).

`pytest` hijacks the loader mechanisms to in order to do its assertion rewriting, and the loader object it returns from `pkgutil.get_loader` does not fulfill the Python API, so the `get_filename` and `get_source` calls will fail.  Had this exception been caught, the function would've gone forward unaffected and the tests would've done their job (and passed).  

Now, since it's not reasonable to write special conditions for an exception caused exclusively by `pytest`, we have to work around this problem indirectly and in less sure-fire ways.  Furthermore, the code is still open to issues of this nature.

Reply:
Ouch. Would it suffice to add `TypeError` to the list?

Reply:
The `TypeError` is for the other situation that uses `inspect.getfile`.

Worse yet, after fixing the `pytest` issue, I hit [this independent `pkgutil.get_loader` issue](https://bugs.python.org/issue14710) in Python 3.4.0.  Thankfully, upgrading to Python 3.4.4 fixes it.